middleware.js

```diff
@@ -1,9 +1,35 @@
 "use strict";
 
+
+const acceptLanguage = require("accept-language");
+
+const { FluentError } = require("./locale-utils");
 const mozlog = require("./log");
 
 
 const log = mozlog("middleware");
+
+
+// adds the request object to a res.local var
+function addRequestToResponse (req, res, next) {
+  res.locals.req = req;
+  next();
+}
+
+
+// picks available language by Accept-Language and assigns to request
+function pickLanguage (req, res, next) {
+  acceptLanguage.languages(req.app.locals.AVAILABLE_LANGUAGES);
+  const pickedLanguage = acceptLanguage.get(req.headers["accept-language"]);
+  req.pickedLanguage = pickedLanguage;
+  const fluentBundle = req.app.locals.FLUENT_BUNDLES[pickedLanguage];
+  req.fluentFormat = (message, args = null, errors = null) => {
+    return fluentBundle.format(fluentBundle.getMessage(message, args));
```

Comment:
This code will work fine in the optimistic scenario of most localizations being complete :) To better support the more realistic scenario of some localizations missing some translations, I'd like to suggest basing this code on the concept of _languages as a list_.

The idea is to negotiate the _list of languages_ requested by the user against the list of available languages. The output is another list. The default language should always be the last item of the resulting list. You can use the `negotiateLanguages` function from the `fluent-langneg` package to perform this negotation reliably.

Unfortunately, AFAICT, the `accept-language` package doesn't expose a method which returns a list of languages parsed from the header. `fluent-langneg` provides a very simple implementation which should be enough for this use-case.

Here's how I'd use it:

```js
import {negotiateLanguages, acceptedLanguages} from "fluent-langneg";

function pickLanguage (req, res, next) {
  // Parse the header into a list of language tags.
  const requested = acceptedLanguages(req.headers["accept-language"]);
  // Negotiate the list of languages for the current user, in order of their preference.
  const supportedLocales = negotiateLanguages(
    requested,
    req.app.locals.AVAILABLE_LANGUAGES
    {defaultLocale: "en-US"}
  );

  req.fluentFormat = (id, args = null, errors = null) => {
    for (let locale of supportedLocales) {
      let bundle = req.app.locals.FLUENT_BUNDLES[locale];
      if (bundle.hasMessage(id)) {
        const message = bundle.getMessage(id);
        return bundle.format(message, args);
      }
    }
    // If no bundle has this message (even the default),
    // fall back to the identifier.
    return id;
  };

  next();
}
```

Reply:
If you don't go this route (although I think you should), note that there's a small bug in the current code. The `args` object should be passed as the second argument to `format` rather than `getMessage`. After all, a single message might be used to produce many different translations in the UI, each for different arguments.

```diff
-return fluentBundle.format(fluentBundle.getMessage(message, args));
+return fluentBundle.format(fluentBundle.getMessage(message), args);
```